### PR TITLE
ADR 773: Add exhaustive test tool for PerseusItem parser

### PR DESCRIPTION
**Pull request by benchristel**
Before we deploy the parsing code described in [ADR 773][1] to production, we
want to make sure it can handle all our production data.

[1]: https://khanacademy.atlassian.net/wiki/spaces/ENG/pages/3318349891/ADR+773+Validate+widget+data+on+input+in+Perseus

This PR adds a tool that will run an exhaustive set of tests, checking the
parser code against every single content item. It tries to summarize the parse
errors intelligently, uniquing them by object path. The output of this tool can
be used to create a suite of regression tests for the parser, which will ensure
it can handle old data formats in perpetuity.

Issue: LEMS-2581

## Test plan:

Run the test tool as described in the comment in
`exhaustive-test-tool/index.ts`. (Note: this will take approximately 4 hours to
run, plus a couple hours to download the data).

You should get 188 errors as measured by `ls /tmp/test-results | wc -l`

### CODE REVIEW

File: packages/perseus/src/util/parse-perseus-json/exhaustive-test-tool/index.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    }
+    if (!Array.isArray(assessmentItems)) {
+        return;
+    }
+
+    for (const rawItem of assessmentItems.map(getAssessmentItemData)) {
+        for (const mismatch of getMismatches(rawItem)) {
+            const desc = describeMismatch(mismatch);
+            const hash = sha256(desc);
+            await fs.mkdir("/tmp/test-results/" + hash, {recursive: true});
```
Comment (benchristel) [inline packages/perseus/src/util/parse-perseus-json/exhaustive-test-tool/index.ts:68]: It's pretty awkward that this output directory is hardcoded. I should make it an argument to the script.

File: packages/perseus/src/util/parse-perseus-json/exhaustive-test-tool/index.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        return "undefined";
+    }
+    if (Array.isArray(value)) {
+        return "array";
+    }
+    return typeof value;
+}
+
+async function* listFilesRecursively(dir: string) {
+    for await (const entry of await fs.opendir(dir, {recursive: true})) {
```
Comment (jeremywiebe) [inline packages/perseus/src/util/parse-perseus-json/exhaustive-test-tool/index.ts:124]: I haven't seen many uses of generator functions outside of examples.  Cool!

**Review by jeremywiebe — APPROVED** [review]
Cool!  I like this approach!

File: packages/perseus/src/util/parse-perseus-json/exhaustive-test-tool/index.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+function describeMismatch(mismatch: Mismatch): string {
+    const path = formatPath(mismatch.path)
+        // Replace numbered widget IDs with generic ones, e.g.
+        // ["interactive-graph 1"] -> ["interactive-graph N"]
+        .replace(/\["([^"\d]+)\d+"]/g, `["$1N"]`)
+        // Replace array indices with generic ones, e.g.
+        // [5] -> [N]
+        .replace(/\[(\d+)]/g, "[N]");
+    return `${path} -- expected ${mismatch.expected.join(", ")}; got ${typeName(mismatch.badValue)}\n`;
```
[the remark this answers is not in the document]
Comment (benchristel) [reply packages/perseus/src/util/parse-perseus-json/exhaustive-test-tool/index.ts:107]: Both issues could be solved at the cost of an additional line:

```
itemData.question.widgets[N].abc
        expected: some value
        received: some other value
```

[the remark this answers is not in the document]
Comment (benchristel) [reply packages/perseus/src/util/parse-perseus-json/exhaustive-test-tool/index.ts:107]: One downside to splitting up the error messages into multiple lines is that Unix tools (grep, sort) operate on lines.

I'd sort of prefer the on-disk format to be one-line-per-record for that reason. It's easy enough to reformat the data with find-replace when I'm creating a TODO list from the error messages.

**Review by catandthemachines — APPROVED** [review]
Looks good to me!

File: packages/perseus/src/util/parse-perseus-json/exhaustive-test-tool/index.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeremywiebe) [reply packages/perseus/src/util/parse-perseus-json/exhaustive-test-tool/index.ts:111]: That's true.  If we want this to be basically a data format, then one-per-line is probably best.

**Pull request merged**